### ISSUE TITLE
Javascript target generated code for Java8.g4 grammar is not valid

### ISSUE THREAD

**Issue by jundl77** [antlr/antlr4#1885]
### Problem
The generated Java 8 lexer contains the following code that is not valid javascript:
(with the [Java8.g4](https://github.com/antlr/grammars-v4/blob/master/java8/Java8.g4) grammar)

```js
Character.toCodePoint((char)_input.LA(-2), (char)_input.LA(-1)));
```

This issue is very similar to similar to #1397. Applying the patch proposed in #1397 has no effect.

### Possible solution?
The ```(char)``` cast is the specific problem. Removing the cast works, however I am not sure if that does not cause other problems. Replacing the char cast with the following also works, however I'm again not sure if that doesn't change the intended behaviour. 

```js
String.fromCharCode(_input.LA(-2));
```

**Comment by BurtHarris** [comment]
Did we get this cleared up @jundl77?   

The problem is that the grammar you refer to is not target-independent.    In particular [line 1745](https://github.com/antlr/grammars-v4/blob/master/java8/Java8.g4#L1745) and similar contain embedded **Java** (not JavaScript) code.   

@part, this Java8 sample seems to cause ongoing confusion.  Its in repository labeled with 'Grammars written for ANTLR v4; expectation that the grammars are free of actions.'   But unfortunately 'free of actions' doesn't mean 'cross-target portable'.

**Comment by KvanTTT** [comment]
Actually, some grammars (PHP, Antlr4, Swift) are not free of actions due to a limitation of context-free languages.

**Comment by BurtHarris** [comment]
@kvanTTT, that may be, but the specific Java grammar lines referenced, are not actions but predicates, and it looks to me like its using a predicate for instead of than expressing something easy to define lexically.    I believe the [relevant definitions for Java 8](http://docs.oracle.com/javase/specs/jls/se8/html/jls-3.html#jls-Identifier) are way simpler than this part of the grammar suggest, saying:

>The "Java letters" include uppercase and lowercase ASCII Latin letters A-Z (\u0041-\u005a), and a-z (\u0061-\u007a), and, for historical reasons, the ASCII underscore (_, or \u005f) and dollar sign ($, or \u0024). The $ sign should be used only in mechanically generated source code or, rarely, to access pre-existing names on legacy systems. 
>
> The "Java digits" include the ASCII digits 0-9 (\u0030-\u0039). 

I think for Java there's no need to get into Unicode surrogate pairs etc.   Identifiers seem to be limited to the 7-bit ASCII subset.

**Comment by BurtHarris** [comment]
@jundl77 have a look at the PR I submitted, I think the updated grammar will eliminate the problem you were seeing.

**Comment by jundl77** [comment]
@BurtHarris I have not yet, I have been very busy these last few days, I'll test it though and get back to you. Thanks for the PR.

**Comment by jundl77** [comment]
@BurtHarris Unfortunately not. Well there is no more invalid Javascript code generated, which I guess is the goal - but for me it still does not terminate using ```antlr4```, and with ```antlr4ts``` a certain ```LiteralContext``` is not defined.

**Comment by cowang** [comment]
@BurtHarris Your quote from the Java 8 Spec above stops too soon. It goes on to say:
> Letters and digits may be drawn from the entire Unicode character set, which supports most writing scripts in use in the world today, including the large sets for Chinese, Japanese, and Korean. This allows programmers to use identifiers in their programs that are written in their native languages.

And above that the Java 8 Spec gives the actual definition of a Java letter:

> A "Java letter" is a character for which the method `Character.isJavaIdentifierStart(int)` returns true.

That is, the Java specification document defines a Java letter to be what an implementation method accepts. Clearly, this violates our sense of what a specification document is supposed to do, but that is the definition. I can understand that during implementation all sorts of corner cases are found that require changes to the specification, but my personal opinion is that the spec writers cheated when they threw up their hands and just wrote that sentence instead of taking the time and effort to document what they want that method to do. I can even accept that sometimes one has to move on instead of making things perfect, but that definition has been in the document for many versions. For instance the Java 1 Spec says:

> A Java letter is a character for which the method `Character.isJavaLetter` (§20.5.17) returns true.

And then it makes clear that other language characters than the Latin/English alphabet are valid by mentioning Greek and Cyrillic. This slip-up should have fixed by now. It's like having the Constitution say that the Presidential Powers are whatever the President does.
<end of rant>

Looking at the documentation for `Character.isJavaIdentifierStart(int)` at [the API](https://docs.oracle.com/javase/8/docs/api/java/lang/Character.html#isJavaIdentifierStart-int-) is not much help either. I suspect that if I knew more about Unicode, it might. Nor does looking at the code. Eventually we get to an implementation of `CharacterData01.isJavaIdentifierStart(int ch)` where we learn that if we check for certain bits in the 32 bit character properties, bits "5, 6, 7 may start a Java identifier". And some parts of the documentation along the chain from  `Character.isJavaIdentifierStart(int)` to `CharacterData01.isJavaIdentifierStart(int ch)` seem to disagree with other parts. `<end rant>`

In summary, it's not easy to describe what we mean by an ordinary character in a way that applies to all the languages of the world. Once upon a time, Terry's Java grammar actually included a list of character ranges that may have come very close to what is needed. Until somebody with a knowledge of Unicode is willing to code up such a list again, my recommendation is to make the change that you propose, but to leave in a comment with the correct code using `Character.isJavaIdentifierStart(int)` and add comments that make it clear that the official definition is the Java method.

Meanwhile I intend to submit a documentation bug report at Java.com.

**Comment by jundl77** [comment]
I think what you proposed here, @cowang would make the most sense:
> If we change the Java 8 grammar in order to allow parsers to be built in other languages, I do recommend that we add a comment with working code for using the two methods and a note that these Java methods are the official definition of Java identifiers.

Like @BurtHarris said, people come here expecting the grammar to be language agnostic, and for somebody like me that does not have huge experience dealing with grammars and parsers it can be confusing.

**Comment by BurtHarris** [comment]
Closing this as its too complicated for a non-Java guy like me.

**Comment by BurtHarris** [comment]
Thanks @cowing for the details.   Java has never been my chosen language, and this aspect of it certainly seems rant-worthy.   Defining the language in terms of a feature in its' runtime seems absurd.  I've closed my  PR without merging.

To further complicate the issue I note that [Unicode Version 10](http://www.unicode.org/versions/Unicode10.0.0/)  just released today, adding over 8,000 code points.

**Comment by jundl77** [comment]
@BurtHarris But would it not make more sense to have the grammar be compliant with different target languages than with all of unicode? Maybe that does not perfectly conform to the Java Spec, but I would argue it is far more useful to parse Java in Javascript or another target language, than being able to handle, and I quote from the new Unicode Version 10:

> Nüshu, used by women in China to write poetry and other discourses until the late twentieth century

or another language. Does the pull request you submitted cover most use-cases?

**Comment by BurtHarris** [comment]
@jundl77, you are probably better qualified to answer this than I.   

I have very little invested in Java,  I figure it's time for me to bow out and let people who care about Java figure that out.   I wrote a tool to port Java to TypeScript, (in PowerShell of all things) but didn't care about formal correctness, it just was meant to assist with porting the ANTLR runtime (where symbols do seem to be restricted to 7-bit ASCII.)   It wasn't perfect, but it didn't have to be since Sam & I always reviewed the results.  I won't be writing another Java based tool using ANTLR, at least not myself.

**Comment by parrt** [comment]
hi guys,There is no good solution here to make it target agnostic. The easiest thing to do is to delineate the characters as I used to have, in order to get the grammar working cross-platform.
